### DISCUSSION TITLE
What does noise characterization look like in Mitiq?

### DISCUSSION THREAD

**Discussion by natestemen** [unitaryfund/mitiq#2193] [Ideas]
If we added some noise characterization in Mitiq, what would it do? How would users want to use this?

The grand goal (IMO) would be to characterize the noise enough to then run some sort of optimization algorithm to determine the best QEM technique for the user to use on their device. The first step of the process is learning _something_ about the noise (anything really!). It's then a separate question as to write the aforementioned optimization algo.

Some basic ideas for an entry point in Mitiq:
- characterize the noise with something like https://www.pygsti.info/ (this is not scalable in the long run, but perhaps useful to people now to learn something about their noise)
- model the noise as pure depolarizing and determine its strength
- model the noise as pure dephasing and determine its strength
- pull down the error rates from IBM (and other providers if they exist)

Lets collect other ideas here, so throw them down (they don't have to be any good 😄).

**Comment by purva-thakre** [comment]
I agree with items 1 and 4. 

Just putting these links here for a later discussion:

https://arxiv.org/abs/2307.08741

https://github.com/tempoCollaboration/OQuPy

**Comment by DustyChalk** [comment]
David Mazziotti described how he took "noiseprints" (noise fingerprints) of quantum computers from within a circuit by running a model.  Perhaps we could either leverage the results of his work, or run simple circuits that exercised typical strings of gates to create our own database of noiseprints.  This was before the days of logical qubits, I have no idea how that would map to logical qubits.
https://arxiv.org/abs/2104.14552

They've identified a previously unknown portion of the causes of noise to cosmic rays:
https://www.newscientist.com/article/2417786-quantum-computers-are-constantly-hampered-by-cosmic-rays/

**Comment by purva-thakre** [comment]
Here are my slides for the first discussion about planning the noise characterization RFC. 

https://docs.google.com/presentation/d/1y94bf3TTCZqm_I3jrFMtYj4P8gR9SGH3pACSDCKwzPk/edit?usp=sharing

**Comment by purva-thakre** [comment]
Link to the literature review document: https://docs.google.com/document/d/1mmjhP6LKcKDegnvmkkxMvbEHXrRlSUXsZ063sqwjFxk/edit?usp=sharing

**Comment by purva-thakre** [comment]
Came across this paper that utilizes surface codes to characterize noise in a superconducting device. https://arxiv.org/abs/2108.10457

A path worth exploring is QEM or a combination of QEM/QEC to characterize device noise which could then be used to improve the performance of QEM and/or QEC.

**Reply by purva-thakre** [reply]
SDPs with tomography, might be worth looking into - https://arxiv.org/abs/2501.18546

**Reply by cosenal** [reply]
oh, very interesting. @vprusso this is our jam!

**Reply by vprusso** [reply]
Our time to shine 🌟 !

**Comment by hrushikesh890** [comment]
As someone actively engaged in this field, I would like to highlight a few noteworthy papers that focus on noise characterization or leverage it to enhance error mitigation:
	1.	QuEst: Graph Transformer for Quantum Circuit Reliability Estimation
([(https://arxiv.org/pdf/2210.16724])
This work employs a noise graph transformer to estimate the reliability and fidelity of quantum circuits.
	2.	Reliability-based Zero Noise Extrapolation (ZNE)
([Paper 1](https://ieeexplore.ieee.org/document/10313621), [Paper 2](https://ieeexplore.ieee.org/document/10821394))
These studies use circuit reliability as the foundation for extrapolation. The reliability is analytically derived from noise characterization data, including gate errors, measurement errors, and thermal relaxation errors.
	3.	Q-BEEP: Quantum Bayesian Error Mitigation Employing Poisson Modeling over the Hamming Spectrum
([Link to paper](https://dl.acm.org/doi/10.1145/3579371.3589043))
This paper uses noise characterization data, such as gate and measurement errors and thermal relaxation errors, to determine Poisson distribution parameters for error mitigation.

I would love to actively contribute to this issue in the future if possible.

**Reply by purva-thakre** [reply]
HI, @hrushikesh890! Thanks for the links to the papers. Right now, this issue is still in the exploration stage. There are a couple of ideas floating around but I am not too sure as to which one will end up being implemented in Mitiq. If you have any ideas at the gate level, feel free to add your comments to this discussion. 

From what I found, there are a multitude of noise characterization techniques. I don't know if we want to add a particular noise characterization technique to mitiq or assume the user already has access to the noise characterization data such that the calibrator module spits out which QEM technique is preferable. 

We also know that noise characterization can be specifically designed for a QEM technique. https://github.com/unitaryfund/mitiq/issues/2197#issuecomment-2477689785
